plugins/usd/module/vtkF3DFaceVaryingPointDispatcher.cxx

```diff
@@ -0,0 +1,148 @@
+
+//------------------------------------------------------------------------------
+vtkF3DFaceVaryingPointDispatcher::~vtkF3DFaceVaryingPointDispatcher() = default;
+
+//------------------------------------------------------------------------------
+int vtkF3DFaceVaryingPointDispatcher::RequestData(vtkInformation* vtkNotUsed(request),
+  vtkInformationVector** inputVector, vtkInformationVector* outputVector)
+{
+  vtkPolyData* input = vtkPolyData::GetData(inputVector[0]->GetInformationObject(0));
+  vtkPolyData* output = vtkPolyData::GetData(outputVector->GetInformationObject(0));
+
+  // early exit if all interpolations are "vertex"
+  vtkPointData* originalPointData = input->GetPointData();
+
+  vtkIdType nbArrays = originalPointData->GetNumberOfArrays();
+
+  bool earlyExit = true;
+
+  for (vtkIdType i = 0; i < nbArrays; i++)
+  {
+    vtkDataArray* originalArray = originalPointData->GetArray(i);
+
+    vtkInformation* info = originalArray->GetInformation();
+    int interpType = info->Get(vtkF3DFaceVaryingPointDispatcher::INTERPOLATION_TYPE());
+
+    if (interpType != 0) // vertex
+    {
+      earlyExit = false;
+      break;
+    }
+  }
+
+  if (earlyExit)
+  {
+    // nothing to do, just return the input
+    output->ShallowCopy(input);
+    return 1;
+  }
+
+  vtkPoints* originalPoints = input->GetPoints();
+  vtkCellArray* originalFaces = input->GetPolys();
+
+  vtkNew<vtkPoints> newPoints;
+  vtkNew<vtkCellArray> newFaces;
+
+  vtkIdType nbCells = originalFaces->GetNumberOfCells();
+  vtkIdType nbConnectivity = originalFaces->GetNumberOfConnectivityIds();
+
+  // the number of output points is the number of total cells connectivity
+  newPoints->SetNumberOfPoints(nbConnectivity);
+
+  vtkPointData* newPointData = output->GetPointData();
+
+  // Use the interpolation type metadata to know if the array can be shallow copied
+  // And initialize the arrays
+  for (vtkIdType i = 0; i < nbArrays; i++)
+  {
+    vtkDataArray* originalArray = originalPointData->GetArray(i);
+
+    vtkInformation* info = originalArray->GetInformation();
+    int interpType = info->Get(vtkF3DFaceVaryingPointDispatcher::INTERPOLATION_TYPE());
+
+    if (interpType == 0) // vertex
+    {
+      vtkDataArray* newArray = vtkDataArray::CreateDataArray(originalArray->GetDataType());
+      newArray->SetNumberOfComponents(originalArray->GetNumberOfComponents());
+      newArray->SetNumberOfTuples(nbConnectivity);
+      newArray->SetName(originalArray->GetName());
+
+      newPointData->AddArray(newArray);
+      newArray->Delete();
+    }
+    else
+    {
+      // if faceVarying, just copy
+      newPointData->AddArray(originalArray);
+    }
+  }
+
+  // copy attribute flags from input
+  for (int attribute : { vtkDataSetAttributes::NORMALS, vtkDataSetAttributes::TCOORDS })
+  {
+    vtkDataArray* array = originalPointData->GetAttribute(attribute);
+    if (array)
+    {
+      newPointData->SetActiveAttribute(array->GetName(), attribute);
+    }
+  }
```

Comment:
unless I'm mistaken, this whole block could be replaced by a shallow copy then the replacement array code only.

Reply:
let me try

Reply:
done